File: src/skills/ref/repo_check.py
"""Repository-level validation helpers for skills/ content."""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Sequence

from .errors import SkillError
from .parser import read_properties
from .prompt import to_prompt
from .validator import validate


def iter_skill_dirs(skills_root: Path) -> list[Path]:
    """Return immediate child directories under the skills root."""
    if not skills_root.is_dir():
        return []

    return sorted(path for path in skills_root.iterdir() if path.is_dir())


def skill_dir_from_path(path: Path, skills_root: Path) -> Path | None:
    """Map a path under skills/ to its top-level skill directory."""
    resolved_path = path.resolve(strict=False)
    resolved_root = skills_root.resolve(strict=False)

    try:
        relative = resolved_path.relative_to(resolved_root)
    except ValueError:
        return None

    if not relative.parts:
        return None

    skill_dir = skills_root / relative.parts[0]
    if skill_dir.exists() and not skill_dir.is_dir():
        return None

    return skill_dir


def resolve_skill_dirs(paths: Sequence[Path], skills_root: Path) -> list[Path]:
    """Resolve changed paths to the set of skill directories to validate."""
    if not paths:
        return iter_skill_dirs(skills_root)

    selected: list[Path] = []
    for path in paths:
        skill_dir = skill_dir_from_path(path, skills_root)
        if skill_dir is not None and skill_dir not in selected:
            selected.append(skill_dir)

    return sorted(selected)


def validate_skill_dirs(skill_dirs: Sequence[Path]) -> list[str]:
    """Validate skill directories and ensure prompt generation still works."""
    errors: list[str] = []

    for skill_dir in skill_dirs:
        problems = validate(skill_dir)
        if problems:
            errors.extend(f"{skill_dir}: {problem}" for problem in problems)
            continue

        try:
            read_properties(skill_dir)
        except SkillError as exc:
            errors.append(f"{skill_dir}: {exc}")

    if errors:
        return errors

    try:
        to_prompt(list(skill_dirs))
    except SkillError as exc:
        errors.append(f"prompt generation failed: {exc}")

    return errors


def main(argv: Sequence[str] | None = None) -> int:
    """Run repository-level skill validation."""
    args = list(argv) if argv is not None else sys.argv[1:]
    skills_root = Path("skills")

    if not skills_root.is_dir():
        print(f"Skills root not found: {skills_root}", file=sys.stderr)
        return 1

    skill_dirs = resolve_skill_dirs([Path(arg) for arg in args], skills_root)
    if not skill_dirs:
        print("No skill directories selected for validation.")
        return 0

    errors = validate_skill_dirs(skill_dirs)
    if errors:
        for error in errors:
            print(error, file=sys.stderr)
        return 1

    print(f"Validated {len(skill_dirs)} skill(s) and generated <available_skills> XML.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/asctl/internal/repocheck/repocheck.go
// Package repocheck orchestrates repository-level skill validation.
package repocheck

import (
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/nq-rdl/agent-skills/tools/asctl/internal/parser"
	"github.com/nq-rdl/agent-skills/tools/asctl/internal/prompt"
	"github.com/nq-rdl/agent-skills/tools/asctl/internal/validator"
)

// IterSkillDirs returns immediate child directories under skillsRoot, sorted.
func IterSkillDirs(skillsRoot string) ([]string, error) {
	entries, err := os.ReadDir(skillsRoot)
	if err != nil {
		return nil, err
	}
	var dirs []string
	for _, e := range entries {
		if e.IsDir() {
			dirs = append(dirs, filepath.Join(skillsRoot, e.Name()))
		}
	}
	sort.Strings(dirs)
	return dirs, nil
}

// SkillDirFromPath maps any path under skillsRoot to its top-level skill directory.
// Returns the skill dir and true on success; "", false if path is outside skillsRoot.
func SkillDirFromPath(path, skillsRoot string) (string, bool) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return "", false
	}
	absRoot, err := filepath.Abs(skillsRoot)
	if err != nil {
		return "", false
	}
	rel, err := filepath.Rel(absRoot, absPath)
	if err != nil {
		return "", false
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return "", false
	}
	parts := strings.SplitN(rel, string(filepath.Separator), 2)
	if len(parts) == 0 || parts[0] == "" || parts[0] == "." {
		return "", false
	}
	skillDir := filepath.Join(skillsRoot, parts[0])
	info, err := os.Stat(skillDir)
	if err != nil || !info.IsDir() {
		return "", false
	}
	return skillDir, true
}

// ResolveSkillDirs resolves changed paths to the set of skill dirs to validate.
// If paths is empty, all skill dirs under skillsRoot are returned.
func ResolveSkillDirs(paths []string, skillsRoot string) ([]string, error) {
	if len(paths) == 0 {
		return IterSkillDirs(skillsRoot)
	}
	seen := make(map[string]bool)
	var selected []string
	for _, p := range paths {
		dir, ok := SkillDirFromPath(p, skillsRoot)
		if ok && !seen[dir] {
			seen[dir] = true
			selected = append(selected, dir)
		}
	}
	sort.Strings(selected)
	return selected, nil
}

// ValidateSkillDirs validates skill directories and ensures prompt generation works.
// Returns a list of error messages; an empty slice means all skills are valid.
func ValidateSkillDirs(skillDirs []string) []string {
	var errors []string
	var validDirs []string

	for _, dir := range skillDirs {
		errs := validator.Validate(dir)
		if len(errs) > 0 {
			for _, e := range errs {
				errors = append(errors, dir+": "+e)
			}
			continue
		}
		if _, err := parser.ReadProperties(dir); err != nil {
			errors = append(errors, dir+": "+err.Error())
			continue
		}
		validDirs = append(validDirs, dir)
	}

	if len(errors) > 0 {
		return errors
	}

	if _, err := prompt.ToPrompt(validDirs); err != nil {
		errors = append(errors, "prompt generation failed: "+err.Error())
	}

	return errors
}
